File: server/app/schemas/learning_path.py
from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class ClassCreate(BaseModel):
    name: str = Field(..., max_length=120)
    description: Optional[str] = None
    grade: Optional[str] = None
    subject: Optional[str] = None
    student_ids: List[UUID] = Field(default_factory=list)


class ClassMemberOut(BaseModel):
    id: UUID
    user_id: UUID
    username: Optional[str] = None
    nickname: Optional[str] = None
    status: str
    joined_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ClassOut(BaseModel):
    id: UUID
    teacher_id: UUID
    name: str
    description: Optional[str] = None
    grade: Optional[str] = None
    subject: Optional[str] = None
    status: str
    created_at: datetime
    member_count: int = 0
    members: List[ClassMemberOut] = Field(default_factory=list)

    class Config:
        from_attributes = True


class LearningPathResourceIn(BaseModel):
    resource_type: str = Field(..., description="bilibili, file, link, text")
    title: Optional[str] = None
    url: Optional[str] = None
    bv_id: Optional[str] = None
    file_id: Optional[UUID] = None
    metadata: Optional[Dict[str, Any]] = None


class LearningPathResourceOut(LearningPathResourceIn):
    id: UUID
    task_id: UUID
    node_id: Optional[UUID] = None

    class Config:
        from_attributes = True


class LearningPathNodeIn(BaseModel):
    key: Optional[str] = None
    title: str
    description: Optional[str] = None
    node_type: str = Field("learning", description="learning, video, reading, practice, submission, checkpoint")
    order_index: int = 0
    estimated_minutes: int = 45
    required: bool = True
    config: Optional[Dict[str, Any]] = None
    resources: List[LearningPathResourceIn] = Field(default_factory=list)


class LearningPathNodeOut(BaseModel):
    id: UUID
    task_id: UUID
    stage_id: Optional[UUID] = None
    key: str
    title: str
    description: Optional[str] = None
    node_type: str
    order_index: int
    estimated_minutes: int
    required: bool
    config: Optional[Dict[str, Any]] = None
    resources: List[LearningPathResourceOut] = Field(default_factory=list)
    progress: Optional[Dict[str, Any]] = None

    class Config:
        from_attributes = True


class LearningPathStageIn(BaseModel):
    title: str
    description: Optional[str] = None
    order_index: int = 0


class LearningPathStageOut(BaseModel):
    id: UUID
    task_id: UUID
    title: str
    description: Optional[str] = None
    order_index: int

    class Config:
        from_attributes = True


class LearningPathEdgeIn(BaseModel):
    source_key: str
    target_key: str


class LearningPathEdgeOut(BaseModel):
    id: UUID
    task_id: UUID
    source_node_id: Optional[UUID] = None
    target_node_id: Optional[UUID] = None
    source_key: str
    target_key: str

    class Config:
        from_attributes = True


class LearningPathGenerateReq(BaseModel):
    goal: str = Field(..., min_length=2)
    planning_text: str = Field(..., min_length=2)


class LearningPathPlanOut(BaseModel):
    stages: List[LearningPathStageIn]
    nodes: List[LearningPathNodeIn]
    edges: List[LearningPathEdgeIn]
    resources: List[LearningPathResourceIn] = Field(default_factory=list)
    summary: str


class LearningPathCreate(BaseModel):
    title: str = Field(..., max_length=255)
    goal: str
    planning_text: Optional[str] = None
    description: Optional[str] = None
    class_id: Optional[UUID] = None
    assignee_ids: List[UUID] = Field(default_factory=list)
    due_date: Optional[datetime] = None
    publish: bool = False
    nodes: Optional[List[LearningPathNodeIn]] = None
    stages: Optional[List[LearningPathStageIn]] = None
    edges: Optional[List[LearningPathEdgeIn]] = None


class LearningPathUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=255)
    goal: Optional[str] = None
    planning_text: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    due_date: Optional[datetime] = None
    nodes: Optional[List[LearningPathNodeIn]] = None
    stages: Optional[List[LearningPathStageIn]] = None
    edges: Optional[List[LearningPathEdgeIn]] = None


class LearningPathTaskOut(BaseModel):
    id: UUID
    creator_id: UUID
    class_id: Optional[UUID] = None
    title: str
    goal: str
    planning_text: Optional[str] = None
    description: Optional[str] = None
    status: str
    due_date: Optional[datetime] = None
    published_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    assignee_count: int = 0
    avg_progress: float = 0.0

    class Config:
        from_attributes = True


class LearningPathDetailOut(BaseModel):
    task: LearningPathTaskOut
    stages: List[LearningPathStageOut]
    nodes: List[LearningPathNodeOut]
    edges: List[LearningPathEdgeOut]
    assignees: List[Dict[str, Any]] = Field(default_factory=list)
    submissions: List[Dict[str, Any]] = Field(default_factory=list)


class LearningNodeSubmitReq(BaseModel):
    content: Optional[str] = None
    attachment_ids: List[UUID] = Field(default_factory=list)
    mark_complete: bool = True


class LearningNodeReviewReq(BaseModel):
    review_status: str = Field("approved", description="approved, rejected, revise")
    score: Optional[float] = Field(None, ge=0, le=100)
    feedback: Optional[str] = None
    follow_up: Optional[str] = None
    reopen_until: Optional[datetime] = None


class InsightEvidenceOut(BaseModel):
    source_type: str
    source_id: Optional[UUID] = None
    student_id: Optional[UUID] = None
    student_name: Optional[str] = None
    content: str
    occurred_at: Optional[datetime] = None


class InsightActionOut(BaseModel):
    action_type: str
    label: str
    payload: Dict[str, Any] = Field(default_factory=dict)


class LearningInsightOut(BaseModel):
    id: UUID
    scope: str
    class_id: Optional[UUID] = None
    student_id: Optional[UUID] = None
    teacher_id: Optional[UUID] = None
    title: str
    insight_type: str
    severity: str
    summary: str
    affected_student_ids: List[UUID] = Field(default_factory=list)
    evidence: List[InsightEvidenceOut] = Field(default_factory=list)
    suggested_actions: List[InsightActionOut] = Field(default_factory=list)
    status: str
    source: str
    source_fingerprint: Optional[str] = None
    generated_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class LearningInsightStatusUpdate(BaseModel):
    status: str = Field(..., description="new, acknowledged, resolved, dismissed")


class ClassOverviewOut(BaseModel):
    class_info: ClassOut
    metrics: Dict[str, Any]
    trend: List[Dict[str, Any]]
    memory_summary: Dict[str, Any]
    insights: List[LearningInsightOut] = Field(default_factory=list)
    attention_students: List[Dict[str, Any]]
    recent_paths: List[LearningPathTaskOut]


class StudentGrowthOverviewOut(BaseModel):
    student_id: UUID
    profile: Dict[str, Any]
    metrics: Dict[str, Any]
    trend: List[Dict[str, Any]]
    learning_paths: List[LearningPathTaskOut]
    memory_cards: List[Dict[str, Any]]
    parent_summary: str

File: web/src/api/modules/learning_path.ts
import request from '../request'

export interface ClassMemberOut {
  id: string
  user_id: string
  username?: string
  nickname?: string
  display_name?: string
  status: string
  joined_at?: string
}

export interface ClassOut {
  id: string
  teacher_id: string
  name: string
  description?: string
  grade?: string
  subject?: string
  status: string
  created_at: string
  member_count: number
  members: ClassMemberOut[]
}

export interface LearningPathResource {
  id?: string
  task_id?: string
  node_id?: string
  resource_type: 'bilibili' | 'file' | 'link' | 'text' | string
  title?: string
  url?: string
  bv_id?: string
  file_id?: string
  metadata?: Record<string, any>
}

export interface LearningPathNode {
  id?: string
  task_id?: string
  stage_id?: string
  key?: string
  title: string
  description?: string
  node_type: 'learning' | 'video' | 'reading' | 'practice' | 'submission' | 'checkpoint' | string
  order_index: number
  estimated_minutes: number
  required: boolean
  config?: Record<string, any>
  resources: LearningPathResource[]
  progress?: {
    id: string
    status: string
    score?: number
    started_at?: string
    submitted_at?: string
    completed_at?: string
  }
}

export interface LearningPathStage {
  id?: string
  task_id?: string
  title: string
  description?: string
  order_index: number
}

export interface LearningPathEdge {
  id?: string
  task_id?: string
  source_node_id?: string
  target_node_id?: string
  source_key: string
  target_key: string
}

export interface LearningPathTaskOut {
  id: string
  creator_id: string
  class_id?: string
  title: string
  goal: string
  planning_text?: string
  description?: string
  status: string
  due_date?: string
  published_at?: string
  created_at: string
  updated_at: string
  assignee_count: number
  avg_progress: number
}

export interface LearningPathPlanOut {
  stages: LearningPathStage[]
  nodes: LearningPathNode[]
  edges: LearningPathEdge[]
  resources: LearningPathResource[]
  summary: string
}

export interface LearningPathDetailOut {
  task: LearningPathTaskOut
  stages: LearningPathStage[]
  nodes: LearningPathNode[]
  edges: LearningPathEdge[]
  assignees: Array<Record<string, any>>
  submissions: Array<Record<string, any>>
}

export interface ClassOverviewOut {
  class_info: ClassOut
  metrics: Record<string, any>
  trend: Array<Record<string, any>>
  memory_summary: Record<string, any>
  insights: LearningInsightOut[]
  attention_students: Array<Record<string, any>>
  recent_paths: LearningPathTaskOut[]
}

export interface InsightEvidenceOut {
  source_type: string
  source_id?: string
  student_id?: string
  student_name?: string
  content: string
  occurred_at?: string
}

export interface InsightActionOut {
  action_type: string
  label: string
  payload: Record<string, any>
}

export interface LearningInsightOut {
  id: string
  scope: string
  class_id?: string
  student_id?: string
  teacher_id?: string
  title: string
  insight_type: string
  severity: 'low' | 'medium' | 'high' | string
  summary: string
  affected_student_ids: string[]
  evidence: InsightEvidenceOut[]
  suggested_actions: InsightActionOut[]
  status: 'new' | 'acknowledged' | 'resolved' | 'dismissed' | string
  source: string
  source_fingerprint?: string
  generated_at?: string
  created_at: string
  updated_at: string
}

export interface StudentGrowthOverviewOut {
  student_id: string
  profile: Record<string, any>
  metrics: Record<string, any>
  trend: Array<Record<string, any>>
  learning_paths: LearningPathTaskOut[]
  memory_cards: Array<Record<string, any>>
  parent_summary: string
}

export const learningPathApi = {
  generatePlan(data: { title?: string; goal: string; planning_text: string }) {
    return request.post('/learning-paths/generate', data)
  },

  createPath(data: {
    title: string
    goal: string
    planning_text?: string
    description?: string
    class_id?: string
    assignee_ids?: string[]
    due_date?: string
    publish?: boolean
    stages?: LearningPathStage[]
    nodes?: LearningPathNode[]
    edges?: LearningPathEdge[]
  }) {
    return request.post('/learning-paths/', data)
  },

  updatePath(taskId: string, data: Partial<{
    title: string
    goal: string
    planning_text: string
    description: string
    status: string
    due_date: string
    stages: LearningPathStage[]
    nodes: LearningPathNode[]
    edges: LearningPathEdge[]
  }>) {
    return request.put(`/learning-paths/${taskId}`, data)
  },

  listTeacherPaths() {
    return request.get('/learning-paths/')
  },

  listStudentPaths() {
    return request.get('/learning-paths/student')
  },

  getPathDetail(taskId: string) {
    return request.get(`/learning-paths/${taskId}`)
  },

  submitNode(taskId: string, nodeId: string, data: { content?: string; attachment_ids?: string[]; mark_complete?: boolean }) {
    return request.post(`/learning-paths/${taskId}/nodes/${nodeId}/submit`, data)
  },

  reviewSubmission(submissionId: string, data: {
    review_status: 'approved' | 'rejected' | 'revise'
    score?: number
    feedback?: string
    follow_up?: string
    reopen_until?: string
  }) {
    return request.post(`/learning-paths/submissions/${submissionId}/review`, data)
  },

  createClass(data: { name: string; description?: string; grade?: string; subject?: string; student_ids?: string[] }) {
    return request.post('/learning-paths/classes', data)
  },

  listClasses() {
    return request.get('/learning-paths/classes/list')
  },

  getClassOverview(classId: string) {
    return request.get(`/learning-paths/classes/${classId}/overview`)
  },

  updateInsightStatus(insightId: string, status: 'new' | 'acknowledged' | 'resolved' | 'dismissed') {
    return request.patch(`/learning-paths/insights/${insightId}/status`, { status })
  },

  getStudentGrowth(studentId: string) {
    return request.get(`/learning-paths/growth/${studentId}`)
  }
}
